### ISSUE TITLE
TraceText Effect

### ISSUE THREAD

**Issue by ProofOfKeags** [fused-effects/fused-effects#400]
It seems odd to me given that fused-effects intends to be a very performant effects library that it chooses to define its trace effects in terms of `String` instead of `Text`. If I **had** to guess, it'd be due to not wanting to carry a dependency on `text`. But given that one of the main things people complain about in Haskell is the ecosystem not supporting `Text` natively, would it be a bad thing to carry that dependency? If so, is there a place for a "dependencies be damned, most people want this use case" version of fused-effects that something like this could go in?

**Comment by patrickt** [comment]
I think we’re anticipating that most people won’t use `Trace` in a critical path, or would use `Writer`. I wish there were some sort of dual to `Data.String.IsString` that provided a `toString` function (`show` would add quotation marks in the case of `String`), as that would solve this handily. I thought there was such an API in `lens`, which would make it an ideal candidate for `fused-effects-lens` (in something like a `Control.Effect.Trace.FromString` API), but I can’t find it now.

**Comment by ProofOfKeags** [comment]
Yeah, I'm leaning towards a church encoded writer that has access to the logging context as an alternative. But for some reason traces seem like the more honest interpretation of what I'd be doing.

Is something like `StrConv` from `protolude` close to what you'd want in the argument? (Dependencies not withstanding)

**Comment by robrix** [comment]
The `Trace` effect here is probably the least well-thought-out part of the library. The use of `String` is honestly just because it’s in the Prelude, and we’ve yet to revisit it.

I’ve been doing some exploration in https://github.com/fused-effects/fused-effects-system and some more in unrelated projects and have mostly come to the conclusion that `Text` is the right choice, but also that the design space deserves a better treatment of its relationship to (among other things) profiling, logging, observability, and the code it’s tracing through. E.g. in a compiler, do you want the trace to relate strongly to the static callstack in the compiler’s code, or to the code you’re compiling? Maybe you want both, for different tasks. Maybe you want both for the same task, with views on either.

Whatever the “right” answer is, it’s probably not going to come quickly, and it’s probably not going to come in this repo; I’d actually prefer to deprecate the built-in `Trace` effect instead of improving it _in situ_.

**Comment by splintersuidman** [comment]
I was recently running into a similar thing, and I was wondering if there is any reason for the `Trace` effect to be monomorphic in its message type? I could for instance imagine a situation where one would like to have a log of some structured message type (e.g. something that serialises to JSON to interface with a web application) with a pure carrier that aggregates the messages in a monoid and an `IO` carrier that writes the messages to a file handle. The `Writer` effect was a bit too strong for my use case, since I didn’t want to bother to keep track of the log in order to implement `listen` (or `look` for `Accum`).

I prototyped something like that in https://github.com/splintersuidman/fused-effects/commit/b7b4d68417f0b412df2a9cc916b39f07771a90b4. I’m not sure it adheres to all of this library’s style choices, so it’s probably not ready to be added as such (and if it should replace the existing `Trace` effect, it would be a breaking change as well). The monoid carrier I wrote is more general than the one for the `Writer` effect, since I wanted to separate the aggregation interpretation and the logging; i.e., one would use `Log LogMessage` effect instead of having to choose the monoid up front and use something like a `Log [LogMessage]` effect.

With such a polymorphic `Log`/`Trace` effect, it would of course be possible for the user to use `Text` messages while this library need not depend on `text`.
